### DISCUSSION TITLE
How to save to continue?

### DISCUSSION THREAD

**Discussion by Dawson867** [ebarellar/LFColoringBook#7] [Q&A]
Hello,
I asked you how to save an existing drawing to a collectionview to continue redrawing it on twitter, can you guide me how to do it here?
Thank you.

**Comment by ebarellar** [comment]
I think what you are trying to achieve is more complex than just "saving"
To add saving functionality you need to do several things

-  Save the currently drawn image

The current drawing is being processed in the frozenContext variable in the ColoringBookView. So you need to ask the coloring book view to give you a drawing to save whenever you are ready to do it

The easiest way is to convert the frozenContext to a CGImage. Check the call to makeImage() in the draw(_ rect: CGRect) method to see how it's done

Then you need a file from that image, preferably a PNG file

You can do it in two ways. Easiest is to just create a UIImage from it and ask for the pngData() method (check the saveImage method in the ViewController to see how it's done). If you want to create it directly from the CGImage check the createMasks method in the ColoringImage struct, the part where CGImageDestinationCreateWithData is called

Now with that file save it somewhere, directly to the documents folder is a good option. Again, check the saveImage method in ViewController to see how it's done

- Now you need to feed the ColoringBookView with both the coloringImage and the image that was drawn

For this, you need to modify the init method of the ColoringBookView, the good news is that it is almost ready for that functionality.
If you store the drawn image in a PNG format, then retrieve the data, pass it in the init method, and just assign it to the savedImage variable (also call the redrawInFrozenContext method after that). With this, not only you get an initial image, but you assure that the undo/redo process respects the initial image

**Reply by Dawson867** [reply]
Thanks for answering!
I solved this problem according to your instructions but according to my understanding. (because I don't understand yet to fix "init" in ColoringBookView )
I do the following:

1. Create New Drawing -> draw - press save and exit
2. Save Image with Core Data (In ColoringBookView)
- root_Image = savedImage.png
- currently_drawn_image = getCurrentImage().pngData()

//==========
public func getCurrentImage() -> UIImage {
        var image = UIImage()
        if let frozenImage = frozenContext.makeImage() {
           image = UIImage(cgImage: frozenImage)
        }
        return image
    }
//=================

In ViewController: I duplicated "CreateColoringBookView(with image:UIImage, optimized:Bool = true)" just to make method call "setSavedImage" at the end - to continue drawing

self.saveImage(image, optimized: optimized)
-> here->: self.coloringBookView?.setSavedImage(currently_drawn_image)
self.checkColoringBookState()

//=========================
func setSavedImage(_ image: UIImage?) {
            guard let currentImage = image?.cgImage else {
                return
            }
            self.savedImage = NSMutableData()
            if let dest = CGImageDestinationCreateWithData(self.savedImage!,
                                                           kUTTypePNG,
                                                           1, nil){
                CGImageDestinationAddImage(dest,
                                           currentImage, nil)
                if CGImageDestinationFinalize(dest){
                    self.strokes.removeAll()
                }
            }
        }
//===================================

After executing, the result is not correct, and here are my results:
1- when taking the drawing being drawn to continue (redisplaying the original image when not drawing)
2- Make 1 stroke -> show undo button
3- Click the undo button once -> return to the previously saved drawing with 4 strokes -> at the same time the redo button is displayed
4- Click redo button - show the original stroke again and hide the redo button
![1 ](https://user-images.githubusercontent.com/89372071/131035926-90b14abe-ad7f-40d4-8db5-6800fa9be5c5.png)
![2 ](https://user-images.githubusercontent.com/89372071/131035935-29aefab3-da94-42f0-92ac-90fcb71afba8.png)
![3 ](https://user-images.githubusercontent.com/89372071/131035943-e9874d6a-64b6-4068-89be-725d83cf2e69.png)
![4 ](https://user-images.githubusercontent.com/89372071/131035951-04e9b59b-a4a0-4277-baf5-32563928156e.png)

**Comment by ebarellar** [comment]
Yes, seems you also need to call redrawInFrozenContext at the end of your setSavedImage to actually draw the image in the savedImage variable from the start

**Reply by Dawson867** [reply]
Thank you very much.

I also tried running it again, calling "redrawInFrozenContext" and not calling "redrawInFrozenContext" but the result is still the same.

There might be something wrong here:

private func reDraw_CreateColoringBookView(with root_Image:UIImage, optimized:Bool = true) {

.....(unchanged).....

DispatchQueue.global(qos: .background).async{
             let coloringImage = ColoringImage(from: root_Image, optimized: optimized)

.....(unchanged)....

DispatchQueue.main.async{
                 messageView.removeFromSuperview()
                 //Create the Coloring Book View
                 self.coloringBookView = ColoringBookView(coloringImage: coloringImage)

..... (unchanged)....

self.saveImage(root_Image, optimized: optimized)
self.coloringBookView?.setSavedImage(currently_drawn_image)
self.checkColoringBookState()
-end-

**Reply by ebarellar** [reply]
Difficult to say without access to the full code. From what I get your saved image is the "1 2 3 4" drawings right?. It seems the savedImage data is stored fine, but there is a problem correctly drawing it from the start, it seems you are only able to see it when the redrawInFrozenContext is called during the redo/undo routine

**Reply by Dawson867** [reply]
Thanks for your enthusiastic support :)

**Reply by nitingohel** [reply]
Hi,

is there possible you add save funtionality and update the repo please.
